library/lua/custom-raw-tokens.lua

```diff
@@ -0,0 +1,243 @@
+--[[
+custom-raw-s
+Allows for reading custom tokens added to raws by mods
+by Tachytaenius (wolfboyft)
+
+Yes, custom raw tokens do quietly print errors into the error log but the error log gets filled with garbage anyway
+]]
+
+local _ENV = mkmodule("custom-raw-tokens")
+
+local eventful = require("plugins.eventful")
+local utils = require("utils")
+
+local customRawTokensCache = {}
+eventful.onUnload["custom-raw-tokens"] = function()
+    customRawTokensCache = {}
+end
+
+-- Not state, defined later:
+local ensureTable, doTable, getSubtype, rawStringsFieldNames
+local getTokenCore, getRaceCasteTokenCore
+
+--[[
+Function signatures:
+getToken(rawStruct, token)
+getToken(rawStructInstance, token)
+getToken(raceDefinition, casteNumber, token)
+getToken(raceDefinition, casteString, token)
+]]
+function getToken(a, b, c)
+    -- Argument processing
+    assert(type(a) == "userdata", "Expected userdata for argument 1 to getToken")
+    if df.is_instance(df.creature_raw, a) then
```

Comment:
factoring the bodies of the if clauses out into separate local functions would make this function much easier to read